pallets/messages/src/migration/v3.rs

```diff
@@ -0,0 +1,229 @@
+/// try to make maximal progress per step, so that the total time it takes to migrate stays low.
+pub struct MigrateV2ToV3<T: Config, W: weights::WeightInfo>(PhantomData<(T, W)>);
+impl<T: Config, W: weights::WeightInfo> SteppedMigration for MigrateV2ToV3<T, W> {
+	// Cursor type for migration. Not used for migration logic itself, other than
+	// for emitting/tracking migration progress/status.
+	// ((tuple of last message processed), cumulative message count)
+	type Cursor = MessagesCursor<T>;
+	// Without the explicit length here the construction of the ID would not be infallible.
+	type Identifier = MigrationId<31>;
+
+	/// The identifier of this migration. Which should be globally unique.
+	fn id() -> Self::Identifier {
+		MigrationId {
+			pallet_id: *b"pallet::messages::migration::v3",
+			version_from: 2,
+			version_to: 3,
+		}
+	}
+
+	/// The actual logic of the migration.
+	///
+	/// This function is called repeatedly until it returns `Ok(None)`, indicating that the
+	/// migration is complete. Ideally, the migration should be designed in such a way that each
+	/// step consumes as much weight as possible.
+	fn step(
+		cursor: Option<Self::Cursor>,
+		meter: &mut WeightMeter,
+	) -> Result<Option<Self::Cursor>, SteppedMigrationError> {
+		#[cfg(feature = "try-runtime")]
+		// Extra check for try-runtime, since pallet-migrations is not seeded with
+		// the completed migrations.
+		if StorageVersion::new(3) <= Pallet::<T>::on_chain_storage_version() {
+			return Ok(None);
+		}
+		let required = W::v2_to_v3_step();
+		// If there is not enough weight for a single step, return an error. This case can be
+		// problematic if it is the first migration that ran in this block. But there is nothing
+		// that we can do about it here.
+		if meter.remaining().any_lt(required) {
+			return Err(SteppedMigrationError::InsufficientWeight { required });
+		}
+
+		let mut step_count = 0u32;
+		let mut iter = MessagesV2::<T>::drain();
+		let mut last_cursor = cursor.unwrap_or((BlockNumberFor::<T>::zero(), 0, 0, 0));
+		let mut messages_remain = true;
+
+		// We loop here to do as much progress as possible per step.
+		loop {
+			if meter.try_consume(required).is_err() {
+				break;
+			}
+
+			// If there's a next item in the iterator, perform the migration.
+			messages_remain = migrate_single_record::<T>(&mut iter, &mut last_cursor);
+			if !messages_remain {
+				break;
+			} else {
+				step_count += 1;
+			}
+		}
+
+		if step_count > 0 {
+			log::info!(target: LOG_TARGET, "Migrated {}{} messages", step_count, if messages_remain { "" } else { " final" });
+		}
+
+		if !messages_remain {
+			v2::DoneV3Migration::<T>::put(true);
+		}
+		Ok(messages_remain.then_some(last_cursor))
+	}
+}
+
+/// Finalize the migration of [`v2::MessagesV2`] map to [`crate::MessagesV3`]
+/// by updating the pallet storage version.
+pub struct FinalizeV3Migration<T: Config, W: weights::WeightInfo>(PhantomData<(T, W)>);
```

Comment:
This second `SteppedMigration` simply bumps the `StorageVersion` in the pallet. It only runs after the first migration is complete. Might seem overkill; technically we could have just bumped this at the end of the main migration, but this provides more tracking via `pallet-migrations` and was a recommended pattern.